Content.Shared/GameObjects/Components/Mobs/SharedActionsComponent.cs

```diff
@@ -0,0 +1,906 @@
+        // We also ensure the values in this dictionary are never empty.
+        [ViewVariables]
+        private Dictionary<EntityUid, Dictionary<ItemActionType, ActionState>> _itemActions =
+            new();
+
+        // this holds cooldowns for revoked item actions that had a cooldown when they were revoked.
+        // they are used to restore cooldowns when the item re-enters inventory
+        // these are not part of component state and thus not synced to the client.
+        // A system runs periodically to evict entries from this when their cooldowns have expired for a long enough time.
+        private Dictionary<(EntityUid item, ItemActionType actionType), (TimeSpan start, TimeSpan end)> _itemActionCooldowns =
+            new();
+
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(ref _innateActions,"innateActions", null);
+        }
+        protected override void Startup()
+        {
+            foreach (var actionType in InnateActions)
+            {
+                Grant(actionType);
+            }
+        }
+
+
+        public override ComponentState GetComponentState()
+        {
+            return new ActionComponentState(_actions, _itemActions);
+        }
+
+        public override void HandleComponentState(ComponentState? curState, ComponentState? nextState)
+        {
+            base.HandleComponentState(curState, nextState);
+
+            if (curState is not ActionComponentState state)
+            {
+                return;
+            }
+            _actions = state.Actions;
+            _itemActions = state.ItemActions;
+        }
+
+        /// <summary>
+        /// Gets the action state associated with the specified action type, if it has been
+        /// granted, has a cooldown, or has been toggled on
+        /// </summary>
+        /// <returns>false if not found for this action type</returns>
+        public bool TryGetActionState(ActionType actionType, out ActionState actionState)
+        {
+            return _actions.TryGetValue(actionType, out actionState);
+        }
+
+        /// <summary>
+        /// Gets the item action states associated with the specified item if any have been granted
+        /// and not yet revoked.
+        /// </summary>
+        /// <returns>false if no states found for this item action type.</returns>
+        public bool TryGetItemActionStates(EntityUid item, [NotNullWhen((true))] out IReadOnlyDictionary<ItemActionType, ActionState>? itemActionStates)
+        {
+            if (_itemActions.TryGetValue(item, out var actualItemActionStates))
+            {
+                itemActionStates = actualItemActionStates;
+                return true;
+            }
+
+            itemActionStates = null;
+            return false;
+        }
+
+        /// <see cref="TryGetItemActionStates"/>
```

Comment:
1. You probably meant `<seealso` instead for references like this.
2. This doesn't actually work since you're not specifying any parameters to resolve the overloads.
![image](https://user-images.githubusercontent.com/8107459/101679083-227f3c00-3a5f-11eb-9935-d4394a83d51c.png)

Reply:
I've fixed this, but apparently it won't let me add the parameters to resolve the overloads in the case of named tuples (affects Cooldown overloads).

Reply:
Named tuples shouldn't affect overload resolution so using unnamed ones for this probably works?